File: websocket-test-server/websocket-server.py
#!/usr/bin/env python3
"""
WebSocket relay server for Swimnerd scoreboard overlays.
Broadcasts messages from the Simulator to all connected OBS overlay sources.

Usage: python3 websocket-server.py
"""

import asyncio
import websockets

PORT = 8080
clients = set()

async def handler(websocket):
    clients.add(websocket)
    print(f"Client connected. Total clients: {len(clients)}")

    try:
        async for message in websocket:
            print(f"Broadcasting message to {len(clients)} clients")
            # Broadcast to all connected clients
            websockets.broadcast(clients, message)
    except websockets.exceptions.ConnectionClosed:
        pass
    finally:
        clients.discard(websocket)
        print(f"Client disconnected. Total clients: {len(clients)}")

async def main():
    print(f"WebSocket server running on ws://localhost:{PORT}")
    print("Waiting for connections from Simulator and OBS overlays...")

    async with websockets.serve(handler, "localhost", PORT):
        await asyncio.Future()  # Run forever

if __name__ == "__main__":
    asyncio.run(main())

File: websocket-server.js
const WebSocket = require('ws');

const PORT = 8080;
const wss = new WebSocket.Server({ port: PORT });

const clients = new Set();

wss.on('connection', (ws) => {
    clients.add(ws);
    console.log(`Client connected. Total clients: ${clients.size}`);

    ws.on('message', (message) => {
        // Broadcast to all connected clients
        const messageStr = message.toString();
        console.log(`Broadcasting message to ${clients.size} clients`);

        clients.forEach((client) => {
            if (client.readyState === WebSocket.OPEN) {
                client.send(messageStr);
            }
        });
    });

    ws.on('close', () => {
        clients.delete(ws);
        console.log(`Client disconnected. Total clients: ${clients.size}`);
    });

    ws.on('error', (error) => {
        console.error('WebSocket error:', error);
        clients.delete(ws);
    });
});

console.log(`WebSocket server running on ws://localhost:${PORT}`);
console.log('Waiting for connections from Simulator and OBS overlays...');
